tests/unit/services/test_k8s.py

```diff
@@ -264,16 +272,126 @@ def test_get_auth_type(
+                    x_cluster_url="https://api.example.com",
+                    x_cluster_certificate_authority_data="abc",
+                    x_k8s_authorization="abc",
+                    x_client_certificate_data="abc",
+                    x_client_key_data=None,
+                    allowed_domains=["example.com"],
+                ),
+                None,
+            ),
+            (
+                "not allowed x_cluster_url domain",
+                K8sAuthHeaders(
+                    x_cluster_url="https://api.custom.com",
+                    x_cluster_certificate_authority_data="abc",
+                    x_k8s_authorization="abc",
+                    x_client_certificate_data="abc",
+                    x_client_key_data=None,
+                    allowed_domains=["example.com"],
+                ),
+                "Cluster URL https://api.custom.com is not allowed.",
+            ),
         ],
     )
-    def test_validate_headers(self, k8s_headers, expected_error_msg):
+    def test_validate_headers(
+        self, description: str, k8s_headers: K8sAuthHeaders, expected_error_msg: str
+    ):
         # given
         if expected_error_msg is not None:
             with pytest.raises(ValueError) as e:
                 k8s_headers.validate_headers()
-            assert expected_error_msg in str(e.value)
+            assert expected_error_msg in str(e.value), description
+        else:
+            assert k8s_headers.validate_headers() is None, description
+
+    @pytest.mark.parametrize(
+        "description, x_cluster_url, allowed_domains, expected_result, expected_error",
+        [
+            ("Allowed domain", "https://api.example.com", ["example.com"], True, None),
+            (
+                "Subdomain allowed",
+                "https://k8s.cluster.example.com",
+                ["example.com"],
+                True,
+                None,
+            ),
+            (
+                "Not allowed domain",
+                "https://api.notallowed.com",
+                ["example.com"],
+                False,
+                None,
+            ),
+            (
+                "Multiple allowed domains, one matches",
+                "https://api.foo.org",
+                ["example.com", "foo.org"],
+                True,
+                None,
+            ),
+            (
+                "No allowed domains (should skip validation and allow)",
+                "https://api.anything.com",
+                [],
+                True,
+                None,
+            ),
+            (
+                "Malformed URL (should raise ValueError)",
+                "not-a-url",
+                ["example.com"],
+                False,
+                ValueError,
+            ),
+            (
+                "Hostname is None (should raise ValueError)",
+                "http://",
+                ["example.com"],
+                False,
+                ValueError,
+            ),
```

Comment:
can you add the following test cases too? The first one is failing. Can you fix it? 
```
            (
                "Domain that contains allowed domain as substring (should be blocked)",
                "https://api.notexample.com",
                ["example.com"],
                False,
                None,
            ),
            ("Exact domain match", "https://example.com", ["example.com"], True, None),
            (
                "Port in URL should be handled correctly",
                "https://api.example.com:8080",
                ["example.com"],
                True,
                None,
            ),
```